protocols/gossipsub/src/protocol.rs

```diff
@@ -267,16 +266,19 @@ impl Encoder for GossipsubCodec {
     type Item = rpc_proto::Rpc;
     type Error = GossipsubHandlerError;
 
-    fn encode(&mut self, item: Self::Item, dst: &mut BytesMut) -> Result<(), Self::Error> {
-        let mut buf = Vec::with_capacity(item.encoded_len());
-
-        item.encode(&mut buf)
-            .expect("Buffer has sufficient capacity");
-
-        // length prefix the protobuf message, ensuring the max limit is not hit
-        self.length_codec
-            .encode(Bytes::from(buf), dst)
-            .map_err(|_| GossipsubHandlerError::MaxTransmissionSize)
+    fn encode(
+        &mut self,
+        item: Self::Item,
+        dst: &mut BytesMut,
+    ) -> Result<(), GossipsubHandlerError> {
+        // TODO: Replace this with a call to the codec::encode
+        let mut codec: prost_codec::Codec<Self::Item> = prost_codec::Codec::new(item.encoded_len());
```

Comment:
You'd want to do this in the constructor of `GossipsubCodec`.

Also the length we pass in here is mostly a DoS attack prevention. It limits how big the buffer is that we will allocate for an _incoming_ message.

This isn't used for encoding messages so passing the encoded length in here doesn't make any sense :)

In any case, this will go away once you move that to the constructor!